### PR TITLE
Make borrow check results available to Clippy

### PR DESCRIPTION
**Pull request by smoelius**
These changes allow Clippy to reconstruct the MIR on which the borrow checker is run, so that Clippy can reproduce its results.

Most of the changes are about making functions public. A notable exception is the addition of the `NllCtxt` structure. The purpose of this struct is to make it easier to call `nll::compute_regions`. IMHO, the introduction of this struct also untangles the code a bit (i.e., helps to expose what depends on what), as an additional benefit.

For context, Clippy now uses a MIR visitor called `PossibleBorrowers`. Intuitively, this visitor produces an approximation of the compiler's borrower check results. But its imprecision can lead to false negatives in certain circumstances (see https://github.com/rust-lang/rust-clippy/pull/9386#issuecomment-1271479113, for example). My previous attempt to make `PossibleBorrowers` more precise did not end well: it had bad worst case complexity, which was [exposed by some in-the-wild programs](https://github.com/rust-lang/rust-clippy/issues/10134).

The reviewer (@Jarcho) and I have discussed ways of improving `PossibleBorrowers`. But one could argue that the "right" approach is to use "ground truth" rather than an approximation, "ground truth" being the results of the compiler's borrow check analysis.

You can see how these changes will be used in https://github.com/rust-lang/rust-clippy/pull/10173. Relevant discussion starts at https://github.com/rust-lang/rust-clippy/pull/10173#issuecomment-1418319870.

### CODE REVIEW

**Comment by oli-obk** [comment]
r? @oli-obk 

I want to make sure this is the right approach, as just exposing some internal APIs is a refactoring hazard

**Comment by oli-obk** [comment]
I read this PR and the clippy PR and got a few questions now:

* is your main problem with the mir building functions that you can't get that MIR anymore because it usually is already stolen?
* is the information you need (region information) computed exactly in the way you need it by borrowck, but then thrown away instead of returned in https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/query/struct.BorrowCheckResult.html?
    * if so, would polonius facts give you all the information that you need, too?

**Comment by smoelius** [comment]
To answer your questions, I need to enumerate a couple of my assumptions.

The way I use these results is to answer questions of the form: are there borrowers of a variable beyond some set of known borrowers? (Please see [here](https://github.com/smoelius/rust-clippy/blob/e8880c16c551f11b55bbf7b82f1eefd443cd9c60/clippy_utils/src/mir/possible_borrower.rs#L130-L158) for my details.)

The assumptions are:
1. Using the region information is a reasonable way to answer such questions.
2. Rerunning `nll::compute_regions` is a reasonable way to obtain the region information.

> * is your main problem with the mir building functions that you can't get that MIR anymore because it usually is already stolen?

Assuming 1 and 2, yes.

> * is the information you need (region information) computed exactly in the way you need it by borrowck, but then thrown away instead of returned in https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/query/struct.BorrowCheckResult.html?

I'm a little unsure about the "exactly" part. But assuming 1, yes.

>   * if so, would polonius facts give you all the information that you need, too?

Maybe? I know very little about Polonius, but there are a couple of things that concern me.

First, the intent of this PR is to give Clippy access to the same borrow check results the compiler uses in deciding whether code is valid. If I understand correctly, Polonius is not currently used in production. So I would be concerned that using it in Clippy could inadvertently cause bugs in Clippy.

A secondary concern is this:
https://github.com/rust-lang/rust/blob/bd4a96a12d0bf6dc12edf20a45df3a33052c9d7d/compiler/rustc_borrowck/src/consumers.rs#L30
I understand the compiler APIs provide no stability guarantees, generally. But I worry about Clippy relying on a feature that is explicitly unstable.

**Comment by smoelius** [comment]
@oli-obk Were there any other questions I could answer on this topic?

As an aside, if you're concerned about exposing internal APIs, another idea could be the following: add a flag to the compiler configuration to clone the various borrow checker structures after they are created, and then expose the cloned structures through a query.

**Comment by oli-obk** [comment]
Sorry, I read that but then must have dismissed the notification before replying

So... about obtaining the right MIR:

My recommendation would be to change `RedundantClone` to run right after borrowck, while the input to borrowck is still available. Basically, you can overwrite the `borrowck` query, run the original query within your overwrite, then run the `RedundantClone` lint on the `promoted_mir` (which isn't stolen at this point). This would also make the redundant clone lint stop using the `optimized_mir`.

For the `Dereferencing` lint it's a bit harder to do the refactoring, but it should work similarly by running it in the borrowck query.

This refactoring seems useful on its own, as it allows clippy to stop using `optimized_mir` and thus get reliable input and potentially even be faster due to less work (I know `optimized_mir` is already cheap in miri due to various flags, but still).


The second part is obtaining the borrowing information you need. Since what you actually want is to access `NllOutput`, not to run `compute_regions`, you could make the `return_body_with_facts` arg of `do_mir_borrowck` be a `Mode` enum instead, that has `Normal, Polonius, Clippy` modes, and in clippy mode it'll return the values you need similarly to the polonius `Option<Box<BodyWithBorrowckFacts<'tcx>>>` return value.


Do you think that plan could work?

**Comment by smoelius** [comment]
> My recommendation would be to change `RedundantClone` to run right after borrowck, while the input to borrowck is still available. Basically, you can overwrite the `borrowck` query, run the original query within your overwrite, then run the `RedundantClone` lint on the `promoted_mir` (which isn't stolen at this point). This would also make the redundant clone lint stop using the `optimized_mir`.

I don't think I understand. When you say "overwrite the `borrowck` query," you mean replace its code?

Are you suggesting to somehow inline `RedundantClone`'s code into the replacement?

(I'm sorry if I'm being dense.)

`RedundantClone` is a late pass lint, which I think means it gets run here: https://github.com/rust-lang/rust/blob/22f247c6f3ed388cb702d01c2ff27da658a8b353/compiler/rustc_interface/src/passes.rs#L841

The most straightforward way I can see realize your idea would be to add something like a "middle" pass that runs just after this: https://github.com/rust-lang/rust/blob/22f247c6f3ed388cb702d01c2ff27da658a8b353/compiler/rustc_interface/src/passes.rs#L786-L788

And then to change `RedundantClone` and `NeedlessBorrow` to be "middle" pass lints.

Do you think that could work?

> The second part is obtaining the borrowing information you need. Since what you actually want is to access `NllOutput`, not to run `compute_regions`, you could make the `return_body_with_facts` arg of `do_mir_borrowck` be a `Mode` enum instead, that has `Normal, Polonius, Clippy` modes, and in clippy mode it'll return the values you need similarly to the polonius `Option<Box<BodyWithBorrowckFacts<'tcx>>>` return value.

I think this part makes sense to me.

**Comment by oli-obk** [comment]
Sorry, I got compiler-driver brain worms. Let me explain from the beginning:

the `mir_borrowck` query is essentially just a function that the compiler calls (and memoizes, and caches to disk, and ...). The way this works is by rustc having a big function pointer table where it looks up what function to call when anyone calls `tcx.mir_borrowck(some_id)`. compiler drivers (like clippy, rustdoc or miri) can overwrite any query's function pointer. For clippy this can be done in https://github.com/rust-lang/rust/blob/01ffa51366d7af655ec879b77fcc566b6fc57d15/src/tools/clippy/src/driver.rs#L131 . For existing examples that overwrite queries you can look at rustdoc: https://github.com/rust-lang/rust/blob/26255186e2e94e0fe62cfd0965662494b6aab27c/src/librustdoc/core.rs#L263

so... basically you make a new query, and within that query you can call `do_mir_borrowck` manually, by basically copying https://github.com/rust-lang/rust/blob/7e2ecb3cd84eae18e838624feeb2e5cf46b9cc57/compiler/rustc_borrowck/src/lib.rs#L136 and using the `Mode::Clippy`.

This way you get the NllOutput and you can then run the logic of RedundantClone::check_fn directly, without needing to build a LintContext or similar. It's a different way of doing lints, for any lints that don't actually visit expressions but only work on MIR, there's no fundamental difference.

There are some things you'll have to do manually, as you don't have a LintContext that gives you a `ParamEnv` or similar, but all of that information can be obtained via `tcx.param_env(def_id)` or similar queries (this is what `LintContext` does internally)

**Comment by smoelius** [comment]
@oli-obk Thanks very much for your patient response. I didn't know it was possible overwrite a query's function pointer.

I think I'm going to have to try implement this idea to understand its implications.

I'll reply once I've had a chance to do so. Thanks again.

**Comment by oli-obk** [comment]
Oh one thing I just remembered: you can do the "run lints in borrowck query" part even without changing borrowck or rustc at all. This may be worthwile on its own and can done entirely within the clippy repo. Should probably bug some other clippy devs about whether that approach seems maintainable and generally good

**Comment by jyn514** [comment]
Hi @smoelius, it's been a while - are you still planning to work on this? Do you know how to make progress? Feel free to ask for help here or [on Zulip](https://rust-lang.zulipchat.com/) if you get stuck.

**Comment by smoelius** [comment]
Hi, @jyn514. Thanks for checking in. Yes, I am still planning to work on this. I have simply had other priorities. I will try to get backs to this soon. Thanks.

**Comment by smoelius** [comment]
@oli-obk I've been trying to implement your suggestion. ([Here](https://github.com/rust-lang/rust-clippy/commit/031f7df384115faa3d4d39531e0e1d6a13dfc1c4) is the latest version of the associated Clippy PR.) However, I've run into some difficulties.

TBH, even if these difficulties can be overcome, I am concerned that writing lints this way is too prohibitive. Is exposing `BodyWithBorrowckFacts` through a query off the table?

The problems I still face include the following:
- Clippy's dogfood test currently fails with "cycle detected" errors involving type checking, borrow checking, and effective visibilities. The following is an example:
  ```
    error[E0391]: cycle detected when computing type of `recursive::<impl at src/recursive.rs:82:1: 82:21>::warnings::{opaque#0}`
    --> src/recursive.rs:117:30
        |
    117 |     pub fn warnings(self) -> impl Iterator<Item = ClippyWarning> {
        |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        |
    note: ...which requires borrow-checking `recursive::<impl at src/recursive.rs:82:1: 82:21>::warnings`...
    --> src/recursive.rs:117:5
        |
    117 |     pub fn warnings(self) -> impl Iterator<Item = ClippyWarning> {
        |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        = note: ...which requires checking effective visibilities...
        = note: ...which again requires computing type of `recursive::<impl at src/recursive.rs:82:1: 82:21>::warnings::{opaque#0}`, completing the cycle
  ```
- It's not clear (to me) how to cache a combined lint pass structure across `mir_borrowck` calls. I'm trying to run the lints that need borrow check information in a combined pass, similar to how existing lints are run. Right now, I'm [reconstructing the `LintPass` objects and the combined pass](https://github.com/rust-lang/rust-clippy/blob/031f7df384115faa3d4d39531e0e1d6a13dfc1c4/src/driver.rs#L190-L198) on each call to `mir_borrowck` (obviously bad). Note that the lint pass members must not outlive `'tcx`, which complicates things. 
- Clippy uses [`enter_lint_attr` and `exit_lint_attr`](https://github.com/rust-lang/rust-clippy/blob/031f7df384115faa3d4d39531e0e1d6a13dfc1c4/clippy_utils/src/msrvs.rs#L133-L143) to determined the MSRV in tests ([example](https://github.com/rust-lang/rust-clippy/blob/031f7df384115faa3d4d39531e0e1d6a13dfc1c4/tests/ui/needless_borrow.rs#L288)). Since `mir_borrowck` is effectively called on a body, it sidesteps those attributes. One could probably walk the enclosing items, etc. to find those attributes, but having two different approaches to finding them seems like a negative.

If exposing `BodyWithBorrowckFacts` through a query is off the table, I am open to suggestions on diagnosing/fixing the above.

**Comment by oli-obk** [comment]
It's not off the table, I would just like to figure out how to ensure we don't end up exposing everything in the compiler that isn't actually used within the compiler. I'd like to have a good way for clippy to just register its very own queries, but that's not possible at present.

Stepping back from the exact implementation: What you want is a way to compute (and obtain!) things from inside the compiler. As a side constraint, we don't want to make the compiler internal APIs more complex.

So, the options I see are:

* what you originally proposed, just returning more things from the `mir_borrowck` query
* what you did now, running lints in the query (bad because it is very different from how all other crates work, and now you have the issue of getting some precomputed data *into* the query)
* unsafe shenanigans like writing the data into a global static hashmap. So you invoke borrowck, throw away the result, but now you can fetch the information from the global static (ugh, that's obviously ugly)
* add a new query to rustc, `clippy_borrowck_results` or something making it obvious that it is not to be used by anyone outside of clippy. I would go as far as to not even give it an implementation at all, and instead only overwrite it in clippy. To ensure you don't duplicate work, you can make clippy's `mir_borrowck` query invoke your new query and just return the subset of data that it would have otherwise computed. (unclear if really better than just sticking the information into the borrowck results like you originally had)

So yea, I'm not sure where to go from here, but I guess as long as clippy is ok with that lint getting removed at a moments notice if borrowck makes a large enough refactoring so that we can't provide this information anymore easily, we can go with your original option.

**Comment by smoelius** [comment]
Thanks very much for your detailed response.

> So yea, I'm not sure where to go from here, but I guess as long as clippy is ok with that lint getting removed at a moments notice if borrowck makes a large enough refactoring so that we can't provide this information anymore easily, we can go with your original option.

May I ask, can you see a point in the future when the borrow checker is more stable?

My impression is that "keeping Clippy working" is a factor influencing API changes today. Put another way, API changes may require changes to a Clippy lint, but the information the lint needs will still be available in some form. "Keeping Clippy working" is one of the reasons why Clippy is mirrored in the Rust repo, for example. (Is this impression wrong?)

Can you see a point in the future when a Clippy lint could rely on borrow checker results, and the same standard (assuming it's correct) could apply?

**Comment by oli-obk** [comment]
> Can you see a point in the future when a Clippy lint could rely on borrow checker results, and the same standard (assuming it's correct) could apply?

yea, I'd assume once we get poloniusification (not switching to, but migrating to a similar scheme) done, we have a certain stability of the output for a decade or so. But that migration may be unduly burdened by having to support clippy. It may be reasonable to port, or it may require rewriting every lint using borrowck results from scratch.

**Comment by smoelius** [comment]
It sounds then like the most reasonable option would be to wait for the Polonius migration to complete, and revisit this issue thereafter. Would you agree?

**Comment by smoelius** [comment]
> Maybe it would be ok for the lint to just not do anything for a bit while it gets patched up, instead of having to fix it immediately when a big refactoring happens?

The lints affected by these changes would `redundant_clone` and `needless_borrow`. Your suggestion is that the Clippy source code somehow include changed and unchanged versions of those lints, so that:
- Effects of borrow checker refactorings on Clippy would be visible.
- Clippy continues to function either way.

Is that right? (Sorry if I'm misunderstanding.)

**Comment by oli-obk** [comment]
Basically we'd just suddenly feed you empty borrowck data. This will break those lints, and it would then be up to the refactoring author and reviewer to decide whether to ask you to fix it or whether they can fix it in a follow up PR. This may end up in stable for a release or two if it can't be done fast enough

**Comment by smoelius** [comment]
@jarcho Can I ask your opinion on having two versions of `redundant_clone` and `needless_borrow` in Clippy, one version like now and one version that uses `borrowck` results?

Based on https://github.com/rust-lang/rust/pull/108328#issuecomment-1623115582 and what follows, I see the implied subgoals as:
- The `borrowck` versions should be built regularly as part of the Rust repo, so that API changes affecting them are visible.
- The `borrowck` versions should be easy to disable, if fixing them would be onerous.
- Clippy's infrastructure should allow testing both versions, since the `borrowck` versions should produce additional warnings.
- Having the two versions should not be detrimental to Clippy's performance.

The easiest way I can see to achieve the above is:
- put the `borrowck` versions behind both a Cargo feature and a `clippy.toml` setting.

Both options would need to be enabled to use the `borrowck` versions of the lints. The Cargo feature would be on-by-default, but could be turned off by Rust developers. The `clippy.toml` setting would be off-by-default, but could be enabled by `ui-toml` tests and users wanting to try out the `borrowck` versions.

What you do think?

(@oli-obk Please correct me if I've misinterpreted or contradicted anything you wrote.)

**Comment by oli-obk** [comment]
I don't think we need cargo features. Just having borrowck return an option that we can turn into None at some point seems sufficient.

**Comment by smoelius** [comment]
> I don't think we need cargo features. Just having borrowck return an option that we can turn into None at some point seems sufficient.

The situation I was imagining was (say) the contents of the `BodyWithBorrowckFacts` struct needs to change, but incorporating those changes into Clippy would be a hassle. A Cargo feature could provide an easy way to just "turn off" that code and avoid any associated compilation problems.

**Comment by oli-obk** [comment]
I expect changes to be either small enough for that to be ok, or massive enough for it to become unrecognizeable. We could keep the old datastructures, so no code breakage, but return `None`.

Though at this point I'm feeling like this puts too much burden on the maintenance side of clippy. I'm sorry for making you go through all these hoops only to end up accepting that your original proposal is probably the best one. Sticking the data you need into a field that rustc sets to `None`, but clippy can safely `unwrap` by overwriting the query. As long as the field is named and documented in ways that make it obvious to rustc internal query invokers that they should not use the field (it's `None` anyway), we can go with that and figure out issues with maintenance when we get there.

**Comment by Jarcho** [comment]
Maintaining two implementations of every lint that uses the borrowck results isn't a great option. Especially as more lints start to make use of it the maintenance burden will become untenable.

Would it be reasonable to commit to a narrow interface wrapping the results? With the current two lints the only thing that needs to be queried is if a local has more than a single borrower as a given location.

**Comment by oli-obk** [comment]
That seems like something any change to the borrow checker could keep around, yes

**Comment by Jarcho** [comment]
That sounds like a good way forward then. As far as deciding whether or not to emit a lint, we should be fine with just three queries:
* Is a local currently borrowed
* Is a local currently mutably borrowed
* Is a local currently borrowed by multiple locals

More might be needed to improve lint messages, but losing anything here temporarily wouldn't be too big of a deal.

**Comment by smoelius** [comment]
@Jarcho You're imagining something like the possible borrower adaptation [I proposed](https://github.com/rust-lang/rust-clippy/blob/031f7df384115faa3d4d39531e0e1d6a13dfc1c4/clippy_utils/src/mir/possible_borrower.rs), but in the compiler instead of in Clippy?

Before going down that path, please let me clear up some confusion I may have caused. When I said "two versions" of those lints, I meant in an intuitive sense, e.g., there would be a handful of locations where those lints would choose between one of two code paths.

To be a little more concrete, I imagined the `clippy.toml` setting doing mainly two things:
- Selecting either the [existing `possible_borrower` implementation](https://github.com/rust-lang/rust-clippy/blob/2153c0fcc893534d61b4d66de80e615e7bff6017/clippy_utils/src/mir/possible_borrower.rs), or one that uses `borrowck` results (the one I linked above).
- Selecting either optimized mir or borrowck mir in `redundant_clone` [here](https://github.com/rust-lang/rust-clippy/blob/2153c0fcc893534d61b4d66de80e615e7bff6017/clippy_lints/src/redundant_clone.rs#L82).

Adding to that, I wouldn't expect there to be much practical difference between having just a `clippy.toml` setting, or having a `clippy.toml` setting and a Cargo feature.

In short, I wouldn't expect Clippy's maintenance burden to be very high, either way.

Does this affect your opinion at all, @Jarcho?

**Comment by Jarcho** [comment]
For the lints that currently exist, the maintenance burden isn't all that high. With that said, both forms would still need to be fully tested. Even with testing it still runs the risk that switching over could still completely break the lints (very unlikely for the current lints, but could be more so for future lints). See #108944 as an example of completely breaking `redundant_clone` in a way that only got detected after the fact.

My main point here is having a limited interface that can, with reasonable ease, also be implemented by future borrowck implementations is better than picking through things on clippy's side.

So the answer to your original question is yes, but limited to the minimum interface clippy actually needs.

**Comment by smoelius** [comment]
> So the answer to your original question is yes, but limited to the minimum interface clippy actually needs.

OK. I'll try to make it happen.

**Comment by Dylan-DPC** [comment]
@smoelius any updates on this?

**Comment by smoelius** [comment]
@Dylan-DPC Not currently. But thank you for the ping. I will try to get back to this soon.

**Comment by smoelius** [comment]
I'm going to close this. I may revisit this at some point in the future.
